src/concurrency.md

```diff
@@ -70,7 +71,102 @@ fn main() {
 }
 ```
 
+[ex-rayon-thumbnails]: #ex-rayon-thumbnails
+<a name="ex-rayon-thumbnails"></a>
+## Generate thumbnails for all .jpg in current directory
 
+[![rayon-badge]][rayon] [![glob-badge]][glob] [![image-badge]][image] [![cat-concurrency-badge]][cat-concurrency] [![cat-filesystem-badge]][cat-filesystem]
+
+This example uses the `glob`, `image`, and `rayon` crates to generate thumbnails for all .jpg files in the current directory. 
+The thumbnails are stored in a new folder called `thumbnails`.
+
+Files are found with [`glob::glob_with`] so we can match case insensitively on both `.jpg` and `.JPG`.
+Images are then resized in parallel with [`par_iter`] using [`DynamicImage::resize`].
+
+
+```rust
+extern crate glob;
+extern crate image;
+extern crate rayon;
+
+use std::path::{Path, PathBuf};
+use std::fs;
+use std::fs::File;
+
+use glob::{glob_with, MatchOptions};
+use image::{FilterType, ImageResult};
+use rayon::prelude::*;
+
+fn main() {
+    let files = find_jpg();
+    println!("Found {} files to convert", files.len());
+    
+    for problem in parallel_resize(&files, Path::new("thumbnails"), 300) {
+        println!("{:?} failed with error: {}", problem.file_path, problem.err);
+    }
+}
+
+/// Find all files that have a `.jpg` extension in the current directory
+fn find_jpg() -> Vec<PathBuf> {
+    let mut paths = Vec::new();
+
+    // use `glob_with()` and MatchOptions so we can match case insensitively to .jpg
+    let options = MatchOptions {
+        case_sensitive: false,
+        require_literal_separator: false,
+        require_literal_leading_dot: false,
+    };
+
+    for entry in glob_with("*.jpg", &options).expect("Failed to read glob pattern") {
+        match entry {
+            Ok(path) => paths.push(path),
+            Err(e) => println!("Error with: {:?}", e),
+        }
+    }
+
+    paths
+}
+
+/// Resize `file` to have a maximum dimension of `longest_edge` and save the resized
+/// image to the `thumb_dir` folder
+fn resize_jpg(file: &Path, thumb_dir: &Path, longest_edge: u32) -> ImageResult<()> {
+    // create the folder if it doesn't already exist
+    fs::create_dir_all(thumb_dir)?;
+
+    let img = image::open(file)?;
+    let img = img.resize(longest_edge, longest_edge, FilterType::Nearest);
+
+    let output_path = thumb_dir.join(file);
+
+    let ref mut fout = File::create(output_path.as_path())?;
+    img.save(fout, image::JPEG)?;
+
+    Ok(())
+}
+
+#[derive(Debug)]
+struct ImageProblem {
+    file_path: PathBuf,
+    err: image::ImageError,
+}
+
+/// Wrapper for `resize_jpg` to resize multiple images at the same time with rayon
+/// Returns a vector of the files that had an error during conversion, along with the actual error
+fn parallel_resize(files: &[PathBuf], thumb_dir: &Path, longest_edge: u32) -> Vec<ImageProblem> {
```

Comment:
I would remove this function and substitute it with a call to
```rust
        files.par_iter()
        .map(|path| {
            resize_jpg(path, thumb_dir, longest_edge)
        })
        .collect();
```
collecting all into a `Result<Vec<_>>` or `Vec<Result<_>>` depending if we want to bail on single error or handle one by one.

Reply:
This is one of the parts I had the most trouble with. The way I did it here was so that I could have some files fail, but still finish most of the files. This way the program would do all it could, and then print out the filenames with their specific errors before exiting.

I guess this is more of a design pattern thing than a `rayon` thing, but I think it's the most important part of the example (handling errors in parallel code), since the other examples don't really deal with things that can error.

I'm open to what people with more experience think is the right approach in this case, as this is some of the first parallel code I've written.

Reply:
Actually there would be no parallel specific pattern here as rayon hides nearly all parallelism details. 

You would handle the errors exactly as if you would in normal iterator based interface. For instance by returning a `Result` with file path from each `map` invocation and then handle each `Result` one by one. You can look at some of the application dev examples like [decompress a tarball while removing a prefix from path example](https://rust-lang-nursery.github.io/rust-cookbook/app.html#decompress-a-tarball-while-removing-a-prefix-from-the-paths).

If you'll have any problems or doubts, do not hesitate to ping me :)

Reply:
@nocduro Do you need any help with finishing this example?